File: python/uiflow2/led/apa102/APA102.py
"""
file     APA102
time     2026-09-16
author
email
license  Apache License 2.0
"""

from periph.connection.spi_auto import SPIConnection
from periph.chips.led.apa102 import APA102Full


class APA102:
    """
    note:
        en: ''
    details:
        color: '#C084FC'
        link: https://github.com/tuhde/Periph
        image: ''
        category: Custom
    example: ''
    """

    def __init__(self, bus: int = 0, device: int = 0, n: int = 1):
        """
        label:
            en: '%1 init SPI bus %2 device %3 n %4'
        params:
            bus:
                name: bus
                type: int
                default: '0'
                field: number
                min: '0'
                max: '7'
            device:
                name: device
                type: int
                default: '0'
                field: number
                min: '0'
                max: '7'
            n:
                name: n
                type: int
                default: '1'
                field: number
                min: '1'
                max: '1024'
        """
        self._driver = APA102Full(SPIConnection(bus=bus, device=device, polarity=0, phase=0, baudrate=1000000), n)

    def fill(self, r: int, g: int, b: int):
        """
        label:
            en: '%1 fill all pixels r %2 g %3 b %4'
        params:
            r:
                name: r
                type: int
                field: number
                min: '0'
                max: '255'
            g:
                name: g
                type: int
                field: number
                min: '0'
                max: '255'
            b:
                name: b
                type: int
                field: number
                min: '0'
                max: '255'
        """
        self._driver.fill(r, g, b)

    def set_pixel(self, index: int, r: int, g: int, b: int, pixel_brightness: int = 31):
        """
        label:
            en: '%1 set pixel index %2 r %3 g %4 b %5 brightness %6'
        params:
            index:
                name: index
                type: int
                field: number
                min: '0'
                max: '1023'
            r:
                name: r
                type: int
                field: number
                min: '0'
                max: '255'
            g:
                name: g
                type: int
                field: number
                min: '0'
                max: '255'
            b:
                name: b
                type: int
                field: number
                min: '0'
                max: '255'
            pixel_brightness:
                name: pixel_brightness
                type: int
                default: '31'
                field: number
                min: '0'
                max: '31'
        """
        self._driver.set_pixel(index, r, g, b, pixel_brightness)

    def show(self):
        """
        label:
            en: '%1 show (transmit buffer)'
        """
        self._driver.show()

    def off(self):
        """
        label:
            en: '%1 off (all pixels)'
        """
        self._driver.off()

    def brightness(self) -> int:
        """
        label:
            en: '%1 brightness (0–255)'
        """
        return self._driver.brightness

    def set_brightness(self, value: int):
        """
        label:
            en: '%1 set brightness %2'
        params:
            value:
                name: value
                type: int
                field: number
                min: '0'
                max: '255'
        """
        self._driver.brightness = value

    def set_pixels(self, colors):
        """
        label:
            en: '%1 set pixels from list %2'
        params:
            colors:
                name: colors
                type: list
        """
        self._driver.set_pixels(colors)

    def rotate(self, steps: int = 1):
        """
        label:
            en: '%1 rotate buffer steps %2'
        params:
            steps:
                name: steps
                type: int
                default: '1'
                field: number
        """
        self._driver.rotate(steps)

    def fill_hsv(self, h: float, s: float, v: float):
        """
        label:
            en: '%1 fill hsv h %2 s %3 v %4'
        params:
            h:
                name: h
                type: float
                field: number
                min: '0.0'
                max: '1.0'
            s:
                name: s
                type: float
                field: number
                min: '0.0'
                max: '1.0'
            v:
                name: v
                type: float
                field: number
                min: '0.0'
                max: '1.0'
        """
        self._driver.fill_hsv(h, s, v)

File: nodejs/packages/node-red-contrib-periph-led/nodes/apa102/apa102.js
'use strict';

module.exports = function(RED) {
    const { SPIConnection } = require('periph/src/connection/spi');
    const { APA102Full }    = require('periph/src/chips/led/apa102');

    function APA102DeviceNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        try {
            node.connection = new SPIConnection(
                parseInt(config.spiBus),
                parseInt(config.spiDevice),
                { mode: 0, maxSpeedHz: 1000000 }
            );
            node.driver = new APA102Full(node.connection, parseInt(config.pixelCount));
            node.driver.brightness = Math.max(0, Math.min(255, parseInt(config.brightness) || 255));
        } catch (e) {
            node.error('APA102 init failed: ' + e.message);
        }
        node.on('close', async function(done) {
            if (node.driver) await node.driver.off();
            if (node.connection) await node.connection.close();
            done();
        });
    }
    RED.nodes.registerType('apa102-device', APA102DeviceNode);

    function APA102WriteNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        node.device = RED.nodes.getNode(config.device);

        node.on('input', async function(msg, send, done) {
            if (!node.device || !node.device.driver) {
                node.error('No APA102 device configured', msg);
                done();
                return;
            }
            const strip = node.device.driver;
            const payload = msg.payload;
            try {
                if (typeof payload === 'object' && payload !== null) {
                    if (payload.command === 'off') {
                        await strip.off();
                    } else if (typeof payload.color === 'string') {
                        const hex = payload.color.replace('#', '');
                        const r = parseInt(hex.slice(0, 2), 16);
                        const g = parseInt(hex.slice(2, 4), 16);
                        const b = parseInt(hex.slice(4, 6), 16);
                        await strip.fill(r, g, b);
                    } else if (Array.isArray(payload.pixels)) {
                        strip.set_pixels(payload.pixels);
                        await strip.show();
                    } else if (typeof payload.pixel === 'number') {
                        strip.set_pixel(payload.pixel,
                            payload.r || 0, payload.g || 0, payload.b || 0,
                            payload.pixel_brightness);
                        await strip.show();
                    } else if (typeof payload.r === 'number' ||
                               typeof payload.g === 'number' ||
                               typeof payload.b === 'number') {
                        await strip.fill(payload.r || 0, payload.g || 0, payload.b || 0,
                            payload.pixel_brightness);
                    }
                }
                done();
            } catch (e) {
                done(e);
            }
        });
    }
    RED.nodes.registerType('periph-apa102', APA102WriteNode);
};